File: 3/create.py
import random

def format_float(num) -> str:
    return f"{num:.2f}".rstrip("0").rstrip(".")

def generate_test_case(n):
    # Generate n random floating point numbers between 0 and 10 with 2 decimal places
    numbers = [round(random.uniform(0, 10), 2) for _ in range(n)]

    # Calculate prefix sums
    prefix_sums = [round(sum(numbers[: i + 1]), 2) for i in range(n)]

    return numbers, prefix_sums


def write_files(n):
    numbers, prefix_sums = generate_test_case(n)

    # Write input file
    with open("random.txt", "w") as f:
        f.write(f"{n}\n")
        f.write(" ".join(map(format_float, numbers)))
        f.write(" \n")

    # Write output file
    with open("random-opt.txt", "w") as f:
        f.write(" ".join(map(format_float, prefix_sums)))
        f.write(" \n")


# Set n manually here
n = 1750  # works for 1750 nums

write_files(n)
print(f"Files 'random.txt' and 'random-opt.txt' have been created with {n} numbers.")

File: 3 Distributed Prefix Sum/create.cpp
#include <fstream>
#include <iomanip>
#include <iostream>
#include <random>
#include <sstream>
#include <vector>

std::vector<double> generate_test_case(int n) {
  std::vector<double> numbers(n);
  std::random_device rd;
  std::mt19937 gen(rd());
  std::uniform_real_distribution<> dis(0.0, 10.0);

  for (int i = 0; i < n; i++) {
    numbers[i] = dis(gen);
  }
  return numbers;
}

std::vector<double> calculate_prefix_sums(const std::vector<double> &numbers) {
  std::vector<double> prefix_sums(numbers.size());
  double sum = 0;
  for (size_t i = 0; i < numbers.size(); i++) {
    sum += numbers[i];
    prefix_sums[i] = sum;
  }
  return prefix_sums;
}

void write_files(int n) {
  auto numbers = generate_test_case(n);

  // Write input file
  std::ofstream numbers_file("random.txt");

  numbers_file << std::setprecision(10);

  numbers_file << n << "\n";
  for (const auto &num : numbers) {
    numbers_file << num << " ";
  }
  numbers_file << "\n";
  numbers_file.close();

  // read from input file
  numbers.clear();
  std::ifstream data_file("random.txt");

  data_file >> n;
  for (int i = 0; i < n; i++) {
    double num;
    data_file >> num;
    numbers.push_back(num);
  }

  // Write output file
  auto prefix_sums = calculate_prefix_sums(numbers);
  std::ofstream output_file("random-opt.txt");

  output_file << std::setprecision(10);

  for (const auto &sum : prefix_sums) {
    output_file << sum << " ";
  }
  output_file << std::endl;
  output_file.close();
}

int main() {
  int n = 100000; // Set n manually here
  write_files(n);
  return 0;
}